src/data/tuple.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2020 Reid Barton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton
+-/
+
+import data.equiv.fin
+
+/-!
+# Finite tuples, implemented as functions on `fin n`.
+
+The main advantage of `tuple` over alternatives like `vector` is that
+extracting coordinates (components) of a `tuple` is just function application.
+
+## Main definitions
+
+* `tuple α n` is by definition the type `fin n → α`. We think of it as
+  the `n`-fold product `αⁿ` and imagine a value `x : tuple α n` as
+  representing a vector `(x 0, ..., x (n-1))`.
+* `tuple.append`, `tuple.left` and `tuple.right` together exhibit the
+  isomorphism `αⁿ⁺ᵐ ≅ αⁿ × αᵐ`.
+* `tuple.snoc`, `tuple.init` and `tuple.last` likewise exhibit the
+  isomorphism `αⁿ⁺¹ ≅ αⁿ × α`.
+-/
+
+/-- A (homogeneous) `n`-tuple `α`s, implemented as a function from `fin n`.
+The `j`th component of such a tuple `x` (where `j : fin n`) is simply `x j`.
+
+This definition is marked `reducible` because we really want to think of
+`x : tuple α n` as a function, while writing the type as `tuple` in order
+to support `.` notation. In particular, this means instances for `ι → α`
+also match `tuple α n` (with `ι = fin n`). -/
+@[reducible] def tuple (α : Type*) (n : ℕ) : Type* := fin n → α
+
+namespace tuple
+
+variables {α : Type*}
+
+def nil : tuple α 0 :=
+fin_zero_elim
+
+instance : unique (tuple α 0) :=
+⟨⟨nil⟩, by { intro x, ext i, exact fin_zero_elim i }⟩
+
+/-- Transport a vector across an equality of dimensions.
+This is implemented without using `eq.rec` so that it will reduce
+when evaluated at a constructor `⟨j, h : j < n'⟩` of `fin n'`. -/
+protected def cast {n n' : ℕ} (h : n = n') : tuple α n → tuple α n' :=
+λ x j, x (j.cast h.symm)
+
+@[simp] lemma cast_app {n n' : ℕ} {h : n = n'} {x : tuple α n} {j : fin n'} :
+  (x.cast h) j = x (j.cast h.symm) :=
+rfl
+
+@[simp] lemma cast_id {n : ℕ} {h : n = n} {x : tuple α n} : x.cast h = x :=
```

Comment:
Perhaps:
```suggestion
@[simp] lemma cast_rfl {n : ℕ} {h : n = n} {x : tuple α n} : x.cast h = x :=
```
since `h` is `rfl` not `id`. Maybe there's already precedent for an `id` name here though.